lib/galaxy/jobs/runners/tes.py

```diff
@@ -0,0 +1,566 @@
+        """
+            Send job-script to TES
+        """
+        log.debug("Sending job script to TES Server")
+        url = f"{master_addr}/v1/tasks"
+        try:
+            req = requests.post(url, json=task)
+            try:
+                job_id = req.json()["id"]
+                return job_id
+            except KeyError:
+                log.error(f"TES Server failed to accept the job {req.json()}")
+        except requests.exceptions:
+            log.error(f"{GENERIC_REMOTE_ERROR} on URL {master_addr}")
+
+    def _get_job(self, master_addr: str, job_id: str, view: str = "MINIMAL"):
+        """
+            Get Job Status from TES Server
+        """
+        log.debug(f"Getting status for job id {job_id}")
+        url = f"{master_addr}/v1/tasks/{str(job_id)}"
+        try:
+            req = requests.get(url, params={'view': view})
+            return req.json()
+        except requests.exceptions:
+            log.error(f"{LOST_REMOTE_ERROR} for job id {job_id}")
+
+    def _cancel_job(self, master_addr: str, job_id: str):
+        """
+            Cancel the Job on TES Server
+        """
+        log.debug(f"Cancelling the job id {job_id}")
+        url = f"{master_addr}/v1/tasks/{job_id}:cancel"
+        try:
+            requests.post(url)
+        except requests.exceptions:
+            log.error(f"{GENERIC_REMOTE_ERROR} for cancellation of job id {job_id}")
+
+    def get_upload_path(self, job_id: str, env=None):
+        """
+            For getting upload URL for staging in the files
+        """
+        if env is None:
+            env = []
+
+        encoded_job_id = self.app.security.encode_id(job_id)
+        job_key = self.app.security.encode_id(job_id, kind="jobs_files")
+        endpoint_base = "%sapi/jobs/%s/files?job_key=%s"
+
+        if self.app.config.nginx_upload_job_files_path:
+            endpoint_base = "%s" + \
+                            self.app.config.nginx_upload_job_files_path + \
+                            "?job_id=%s&job_key=%s"
+
+        files_endpoint = endpoint_base % (
+            self.galaxy_url,
+            encoded_job_id,
+            job_key
+        )
+
+        get_client_kwds = dict(
+            job_id=str(job_id),
+            files_endpoint=files_endpoint,
+            env=env
+        )
+        return get_client_kwds
+
+    def find_referenced_subfiles(self, directory: str, command_line: str, extra_files: list):
+        """
+        Return list of files below specified `directory` in job inputs. Could
+        use more sophisticated logic (match quotes to handle spaces, handle
+        subdirectories, etc...).
+        **Parameters**
+        directory : str
+            Full path to directory to search.
+        """
+        if directory is None:
+            return []
+
+        pattern = r'''[\'\"]?(%s%s[^\s\'\"]+)[\'\"]?''' % (escape(directory), escape(os.sep))
+        return self.find_pattern_references(pattern, command_line, extra_files)
```

Comment:
I appreciate this as a proof of concept - it is very impressive - but there are big pieces of code that just duplicate important and well tested parts of the Pulsar client. I'm really hesitant to merge this duplicated functionality, the duplicated code even more so I think.

We're going to push this off at least one release but ideally I would like to take some time and really deeply integrate this with Pulsar.

Reply:
Hi, @jmchilton Thanks for the review. Do you mean to integrate the `find_referenced_subfiles` version as used in tes runner to pulsar to reduce the redundancy?